### PR TITLE
Recursor: Add `--config[=check|=diff|=default]`

### PR DESCRIPTION
**Pull request by fredmorcos**
### Short description
Add `--config[=check|=diff|=default]` to the recursor

### Checklist
I have:
- [X] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [X] compiled this code
- [X] tested this code
- [ ] included documentation (including possible behaviour changes)
- [ ] documented the code
- [ ] added or modified regression test(s)
- [ ] added or modified unit test(s)

### CODE REVIEW

File: pdns/recursordist/rec-main.cc [hunk 1]
```diff
@@ -2893,6 +2894,35 @@ int main(int argc, char** argv)
...
+          exit(0);
+        }
+        catch (const ArgException& argException) {
+          SLOG(g_log << Logger::Warning << "Unable to parse configuration file '" << configname << "'" << endl,
+               startupLog->error("Cannot parse configuration", "Unable to parse configuration file", "config_file", Logging::Loggable(configname)));
+          exit(1);
+        }
+      }
+      else {
+        cout << ::arg().configstring(true, true);
```
Comment (rgacogne) [inline pdns/recursordist/rec-main.cc:2921]: Is there an actual difference between `--config` and `--config=default` at this point?

Comment (fredmorcos) [reply pdns/recursordist/rec-main.cc:2921]: Right... Probably should add a call to `::arg().laxFile(...)` before the block to have `--config` dump a loaded configuration file (with warnings about unknown settings). I'm not sure if it's the right way to do it though...

File: pdns/recursordist/rec-main.cc [hunk 2]
```diff
@@ -2893,6 +2894,35 @@ int main(int argc, char** argv)
...
+                 startupLog->error("No such file", "Unable to open configuration file", "config_file", Logging::Loggable(configname)));
+            exit(1);
+          }
+
+          ::arg().parse(argc, argv);
+          exit(0);
+        }
+        catch (const ArgException& argException) {
+          SLOG(g_log << Logger::Warning << "Unable to parse configuration file '" << configname << "'" << endl,
+               startupLog->error("Cannot parse configuration", "Unable to parse configuration file", "config_file", Logging::Loggable(configname)));
```
Comment (omoerbeek) [inline pdns/recursordist/rec-main.cc:2916]: With a wrong config file I now see:
```
Sep 07 09:33:17 msg="Unable to parse configuration file" error="Cannot parse configuration" subsystem="config" level="0" tid="0" ts="1662535997.193" config_file="./recursor.conf"
```
The ectual underlying `argException.reason` is not included in the error message.

File: pdns/recursordist/rec-main.cc (hunk 1 above)
Comment (omoerbeek) [reply pdns/recursordist/rec-main.cc:2921]: Did you take a look how auth does it?

**Comment by fredmorcos** [comment]
When approving, please don't merge this just yet, I need to run a fixup of the commits.

File: pdns/recursordist/rec-main.cc [hunk 3]
```diff
@@ -2893,9 +2894,43 @@ int main(int argc, char** argv)
...
+          ::arg().parse(argc, argv);
+          exit(0);
+        }
+        catch (const ArgException& argException) {
+          SLOG(g_log << Logger::Warning << "Unable to parse configuration file '" << configname << "': " << argException.reason << endl,
+               startupLog->error("Cannot parse configuration", "Unable to parse configuration file", "config_file", Logging::Loggable(configname), "reason", Logging::Loggable(argException.reason)));
+          exit(1);
+        }
+      }
+      else {
```
Comment (omoerbeek) [inline pdns/recursordist/rec-main.cc:2919]: This is for the `--confg=xyz` (where `xyz` may be empty) case. The man page change does not describe this.
While looking how auth describes this I see it also has `diff`, which would be really nice to have in rec as well 

So can you take the man pages lines from auth and see if `diff` can be added as well?

**Review by omoerbeek — CHANGES_REQUESTED** [review]

File: pdns/recursordist/rec-main.cc (hunk 3 above)
Comment (fredmorcos) [reply pdns/recursordist/rec-main.cc:2919]: Alright

File: pdns/recursordist/docs/manpages/pdns_recursor.1.rst [hunk 4]
```diff
@@ -46,6 +46,11 @@ at `<https://doc.powerdns.com/>`
     chroot the process to *directory*.
 --client-tcp-timeout=<num>
     Timeout in seconds when talking to TCP clients.
+--config
+    Show the current configuration. There are two optional values:
```
Comment (omoerbeek) [inline pdns/recursordist/docs/manpages/pdns_recursor.1.rst:50]:
```suggestion
    Show the current configuration. There are three optional values:
```

File: pdns/recursordist/rec-main.cc (hunk 3 above)
Comment (fredmorcos) [reply pdns/recursordist/rec-main.cc:2919]: Is it better now?

File: pdns/recursordist/rec-main.cc [hunk 5]
```diff
@@ -2893,9 +2894,49 @@ int main(int argc, char** argv)
...
+        if (!::arg().laxFile(configname.c_str())) {
+          SLOG(g_log << Logger::Warning << "Unable to open configuration file '" << configname << "'" << endl,
+               startupLog->error("No such file", "Unable to open configuration file", "config_file", Logging::Loggable(configname)));
+          exit(1);
+        }
+        ::arg().laxParse(argc, argv);
+
+        if (config == "diff") {
```
Comment (omoerbeek) [inline pdns/recursordist/rec-main.cc:2927]: We might want to also check before producing a diff. That would mean the test above becomes "check" or "diff".

**Review by omoerbeek — APPROVED** [review]
Two small nits

File: pdns/recursordist/rec-main.cc [hunk 6]
```diff
@@ -2893,9 +2894,49 @@ int main(int argc, char** argv)
...
+          ::arg().parse(argc, argv);
+          exit(0);
+        }
+        catch (const ArgException& argException) {
+          SLOG(g_log << Logger::Warning << "Unable to parse configuration file '" << configname << "': " << argException.reason << endl,
+               startupLog->error("Cannot parse configuration", "Unable to parse configuration file", "config_file", Logging::Loggable(configname), "reason", Logging::Loggable(argException.reason)));
+          exit(1);
+        }
+      }
+      else if (config == "test" || config == "diff") {
```
Comment (omoerbeek) [inline pdns/recursordist/rec-main.cc:2919]: My suggestion for check before diff was not right, sorry. You will need to restructure the `if ... else` blocks a bit to check check before diff working..

**Review by omoerbeek — CHANGES_REQUESTED** [review]

**Review by omoerbeek — APPROVED** [review]

**Pull request merged**